gemini/use-cases/retail/product_recognition_from_image.ipynb

```diff
@@ -0,0 +1,506 @@
+      "metadata": {
+        "id": "v1RqPUeSk0kt"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Install gcp aiplatform package\n",
+        "! pip3 install --upgrade --user --quiet google-cloud-aiplatform"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "95IwjOvWk_Lc",
+      "metadata": {
+        "id": "95IwjOvWk_Lc"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Restart the kernel after installing package.\n",
+        "import sys\n",
+        "\n",
+        "if \"google.colab\" in sys.modules:\n",
+        "    import IPython\n",
+        "\n",
+        "    app = IPython.Application.instance()\n",
+        "    app.kernel.do_shutdown(True)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "ot7iCovUlEK6",
+      "metadata": {
+        "id": "ot7iCovUlEK6"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Authetication\n",
+        "import sys\n",
+        "\n",
+        "if \"google.colab\" in sys.modules:\n",
+        "    from google.colab import auth\n",
+        "\n",
+        "    auth.authenticate_user()"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "fL3GwbKKlEtE",
+      "metadata": {
+        "id": "fL3GwbKKlEtE"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Setup Vertex project.\n",
+        "PROJECT_ID = \"[your-project-id]\"  # @param {type:\"string\"}\n",
+        "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+        "\n",
+        "import vertexai\n",
+        "\n",
+        "vertexai.init(project=PROJECT_ID, location=LOCATION)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "fHVclx2ElUEF",
+      "metadata": {
+        "id": "fHVclx2ElUEF"
+      },
+      "source": [
+        "# Product Recognition using Gemini Demo"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 9,
+      "id": "xxlXl4hiDKPeITgNTUBMGAFP",
+      "metadata": {
+        "cellView": "form",
+        "id": "xxlXl4hiDKPeITgNTUBMGAFP"
+      },
+      "outputs": [],
+      "source": [
+        "from io import BytesIO\n",
+        "\n",
+        "from IPython.display import display\n",
+        "from PIL import Image\n",
+        "from googleapiclient.discovery import build  # For google custom search engine.\n",
+        "\n",
+        "# @title [Library] Gemini based Product Image Search/Recognition.\n",
+        "import ipywidgets as widgets\n",
+        "import requests\n",
+        "from vertexai.preview.generative_models import GenerationConfig, GenerativeModel, Part\n",
+        "\n",
+        "# Google Custom search engine developer key.\n",
+        "# To get the key, follow the instructions at\n",
+        "# https://developers.google.com/custom-search/v1/overview\n",
+        "YOUR_DEVELOPER_KEY = \"\"  # @param {\"type\": \"string\"}\n",
```

Comment:
I think it's recommended now to use Vertex AI Search instead of Google Custom Search Engine

Migration Guide: https://cloud.google.com/generative-ai-app-builder/docs/migrate-from-cse